[exporter/elasticsearch] remove dedup config

axw opened the pull request:
**Description:**

Remove the `dedup` configuration setting, and always de-duplicate. Elasticsearch does not permit duplicate keys in JSON objects, and this configuration is adding more complexity to the code than it's worth.

I've simplified the `internal/objmodel` API slightly, unexporting the `Sort` methods, which are internally called by the now unconditional call to `Dedup`.

**Link to tracking Issue:**

Closes https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/33773

**Testing:**

Ran the unit tests, which cover deduplication. None of the tests in package elasticsearchexporter covered `dedup: false`.

**Documentation:** <Describe the documentation added.>

Commit:
[exporter/elasticsearch] remove dedup config

Remove the dedup configuration setting, and always deduplicate.
Elasticsearch does not permit duplicate keys in JSON objects,
and this configuration is adding more complexity to the code
than it's worth.

andrzej-stencel commented on exporter/elasticsearchexporter/README.md:
```diff
@@ -127,10 +127,6 @@ behaviours, which may be configured throug the following settings:
              field names for span events. 
   - `fields` (optional): Configure additional fields mappings.
   - `file` (optional): Read additional field mappings from the provided YAML file.
-  - `dedup` (default=true): Try to find and remove duplicate fields/attributes
```
We should not remove a configuration property without first deprecating it, given that this component is in `Beta` stability. Here's the documentation: https://github.com/open-telemetry/opentelemetry-collector/blob/main/CONTRIBUTING.md#beta-components.

axw replied:
I'll create a separate PR tomorrow with just the deprecations. I did it this way because the config does not actually have any functional benefit. Elasticsearch will reject docs with duplicate object keys, so if you disable it then the exporter won't actually be useful.

andrzej-stencel commented:
@axw  Can you split the `dedot` change into a separate PR please?

carsonip commented on .chloggen/elasticsearch-always-dedup.yaml:
```diff
@@ -0,0 +1,29 @@
+change_type: breaking
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: exporter/elasticsearch
```
```suggestion
component: elasticsearchexporter
```

axw replied:
Oops, thanks.

carsonip reviewed (commented):
one nit otherwise lgtm

axw committed:
Update .chloggen/elasticsearch-always-dedup.yaml

Co-authored-by: Carson Ip <carsonip@users.noreply.github.com>

Commit:
Merge branch 'main' into elasticsearch-always-dedup

andrzej-stencel reviewed (changes requested):
According to the [docs](https://github.com/open-telemetry/opentelemetry-collector/blob/v0.104.0/CONTRIBUTING.md#beta-components):

> the option (...) MAY be removed after N+2 or 6 months, whichever comes later

I believe this means we should not remove this option before January 2025 (six months after v0.104.0). We can hide the option behind a feature flag if it helps (not sure if it makes sense in this case?).

You can still make the feature non-operational according to this:

> the option MAY be made non operational already by the same version where it is deprecated

Commit:
Reinstate "dedup" config, make it a no-op

It's against policy to remove within 6 months
of deprecation, but it can be made a no-op.

Commit:
Merge branch 'main' into elasticsearch-always-dedup

axw commented:
@andrzej-stencel thanks for pointing out the docs. I've reinstated the config, but made it a no-op. I didn't add it back into the README though - let me know if you disagree with that.

> the option and the WARN level message MAY be removed after N+2 or 6 months, whichever comes later

I find the "MAY" in this to be a little bit ambiguous. Perhaps it would be worth rewording to:

"the option and WARN level message MUST NOT be removed earlier than N+2 o 6 months, whichever comes later"

WDYT?

andrzej-stencel commented:
> I didn't add it back into the README though - let me know if you disagree with that.

I think if the option exists, it should be described in the README - saying it is deprecated and non-functional.

> I find the "MAY" in this to be a little bit ambiguous. Perhaps it would be worth rewording to:
> "the option and WARN level message MUST NOT be removed earlier than N+2 o 6 months, whichever comes later"

I agree that the wording might be a bit misleading. Your proposal seems OK, please send a PR to update the wording, let's see what other maintainers think.

andrzej-stencel commented on .chloggen/elasticsearch-always-dedup.yaml:
```diff
@@ -0,0 +1,29 @@
+component: elasticsearchexporter
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Disable "dedup" configuration, always de-duplicate.
```
```suggestion
note: Make "dedup" option no-op, always de-duplicate.
```

andrzej-stencel commented on exporter/elasticsearchexporter/README.md:
```diff
@@ -129,10 +129,6 @@ behaviours, which may be configured through the following settings:
     - `raw`: Omit the `Attributes.` string prefixed to field names for log and 
              span attributes as well as omit the `Events.` string prefixed to
              field names for span events. 
-  - `dedup` (default=true; DEPRECATED, in future deduplication will always be enabled):
```
Let's not remove the option from the docs. Let's update the description to say it doesn't do anything.

axw replied:
Will do.

andrzej-stencel commented on exporter/elasticsearchexporter/config.go:
```diff
@@ -160,12 +160,12 @@ type MappingsSettings struct {
 	Mode string `mapstructure:"mode"`
 
-	// Try to find and remove duplicate fields
+	// Dedup is non-operational, and will be removed from January 2025.
```
Maybe let's not commit to dates?
```suggestion
	// Dedup is non-operational, and will be removed in the future.
```

axw replied:
The intention here is to indicate the earliest time at which the config will be removed. I'm not too bothered though.

andrzej-stencel commented on exporter/elasticsearchexporter/config.go:
```diff
@@ -322,8 +322,8 @@ func (cfg *Config) MappingMode() MappingMode {
-	if !cfg.Mapping.Dedup {
-		logger.Warn("dedup has been deprecated, and will always be enabled in future")
+	if cfg.Mapping.Dedup != nil {
+		logger.Warn("dedup has been deprecated, and is always enabled")
```
```suggestion
		logger.Warn("dedup is deprecated, and is always enabled")
```

axw committed:
Apply suggestions from code review

Co-authored-by: Andrzej Stencel <andrzej.stencel@elastic.co>

Commit:
Update test

Commit:
Add mention of "dedup" config back into README

Commit:
Merge branch 'main' into elasticsearch-always-dedup

Commit:
Merge branch 'main' into elasticsearch-always-dedup

Commit:
Merge branch 'main' into elasticsearch-always-dedup

Commit:
Merge branch 'main' into elasticsearch-always-dedup
